File: Klarna.Rest/Klarna.Rest/Models/EMD/ExtraMerchantData.cs
#region Copyright Header
//-----------------------------------------------------------------------
// <copyright file="ExtraMerchantData.cs" company="Klarna AB">
//     Copyright 2015 Klarna AB
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// </copyright>
//-----------------------------------------------------------------------
#endregion
namespace Klarna.Rest.Models.EMD
{
    using System;
    using System.Collections.Generic;
    using Newtonsoft.Json;
    using Newtonsoft.Json.Converters;

    /// <summary>
    /// The model for the EMD attachment.
    /// </summary>
    public class ExtraMerchantData
    {
        /// <summary>
        /// Format used for formatting dates.
        /// </summary>
        private const string DateTimeFormat = "yyyy'-'MM'-'dd'T'HH':'mm':'ss'Z'";

        /// <summary>
        /// Content type.
        /// </summary>
        private const string Type = "application/vnd.klarna.internal.emd-v2+json";

        /// <summary>
        /// Gets the content type.
        /// </summary>
        public string ContentType
        {
            get
            {
                return Type;
            }
        }

        /// <summary>
        /// Gets or sets the body model.
        /// </summary>
        public ExtraMerchantDataBody Body { get; set; }

        /// <summary>
        /// Cast extra merchant data to an attachment.
        /// </summary>
        /// <param name="emd">The extra merchant data to cast.</param>
        public static implicit operator Attachment(ExtraMerchantData emd)
        {
            return new Attachment
            {
                ContentType = emd.ContentType,
                Body = JsonConvert.SerializeObject(
                    emd.Body,
                    new JsonSerializerSettings
                    {
                        NullValueHandling = NullValueHandling.Ignore,
                        DefaultValueHandling = DefaultValueHandling.Ignore,
                        Converters = new List<JsonConverter>()
                        {
                            new IsoDateTimeConverter
                            {
                                DateTimeFormat = DateTimeFormat
                            }
                        }
                    })
            };
        }

        /// <summary>
        /// Cast an attachment to extra merchant data.
        /// </summary>
        /// <param name="attachment">Attachment to cast.</param>
        public static implicit operator ExtraMerchantData(Attachment attachment)
        {
            if (string.IsNullOrEmpty(attachment.ContentType) || !attachment.ContentType.Equals(Type))
            {
                throw new InvalidCastException("Incorrect Content-Type");
            }

            return new ExtraMerchantData
            {
                Body = JsonConvert.DeserializeObject<ExtraMerchantDataBody>(
                    attachment.Body,
                    new IsoDateTimeConverter
                    {
                        DateTimeFormat = DateTimeFormat
                    })
            };
        }
    }
}

File: src/main/java/com/klarna/rest/api/model/emd/ExtraMerchantData.java
/*
 * Copyright 2015 Klarna AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.klarna.rest.api.model.emd;

import com.klarna.rest.api.ObjectMapperProvider;
import com.klarna.rest.api.model.Attachment;
import com.klarna.rest.api.model.Model;
import org.codehaus.jackson.map.ObjectMapper;

import java.io.IOException;

/**
 * The model for the EMD attachment.
 */
public class ExtraMerchantData extends Model {
    /**
     * Extra merchant data content type.
     */
    private static String contentType =
            "application/vnd.klarna.internal.emd-v2+json";

    /**
     * Extra merchant data body model.
     */
    private ExtraMerchantDataBody body;

    /**
     * Create a ExtraMerchantData instance from an Attachment.
     *
     * @param attachment Attachment to convert.
     * @return Parsed ExtraMerchantData instance.
     * @throws IOException For invalid content type or invalid body string.
     */
    public static ExtraMerchantData fromAttachment(
            final Attachment attachment) throws IOException {
        if (attachment.getContentType() == null
                || !attachment.getContentType().equals(contentType)) {
            throw new IOException("Incorrect content type");
        }

        ObjectMapperProvider omp = new ObjectMapperProvider();
        ObjectMapper om = omp.getContext(ExtraMerchantData.class);

        ExtraMerchantDataBody body = om.readValue(
                attachment.getBody(), ExtraMerchantDataBody.class);

        return new ExtraMerchantData().setBody(body);
    }

    /**
     * Get the content type.
     *
     * @return Content type.
     */
    public String getContentType() {
        return contentType;
    }

    /**
     * Get the extra merchant data body model.
     *
     * @return Extra merchant data body model.
     */
    public ExtraMerchantDataBody getBody() {
        return this.body;
    }

    /**
     * Set the extra merchant data body model.
     *
     * @param body Extra merchant data body model.
     * @return Same instance.
     */
    public ExtraMerchantData setBody(final ExtraMerchantDataBody body) {
        this.body = body;

        return this;
    }

    /**
     * Convert this instance to an attachment.
     *
     * @return Attachment.
     * @throws IOException If the body could not be serialized.
     */
    public Attachment toAttachment() throws IOException {
        ObjectMapperProvider omp = new ObjectMapperProvider();
        ObjectMapper om = omp.getContext(ExtraMerchantData.class);

        return new Attachment()
                .setContentType(this.getContentType())
                .setBody(om.writeValueAsString(this.getBody()));
    }
}
